documentation/doxygen.py

```diff
@@ -1992,39 +1992,69 @@ def parse_func(state: State, element: ET.Element):
     func.brief = parse_desc(state, element.find('briefdescription'))
     func.description, templates, params, func.return_value, func.return_values, func.exceptions, search_keywords, func.deprecated, func.since = parse_func_desc(state, element)
 
-    # Friend functions have friend as type. That's just awful. COME ON.
-    if func.type.startswith('friend '):
-        func.type = func.type[7:]
-
     def is_identifier(a): return a == '_' or a.isalnum()
 
     # Extract function signature to prefix, suffix and various flags. Important
     # things affecting caller such as static or const (and rvalue overloads)
     # are put into signature prefix/suffix, other things to various is_*
     # properties.
-    if func.type == 'constexpr': # Constructors
-        func.type = ''
-        func.is_constexpr = True
-    elif func.type.startswith('constexpr'):
-        func.type = func.type[10:]
-        func.is_constexpr = True
-    # For some effing reason, when a constexpr function has decltype(auto)
-    # return type, Doxygen swaps the order of those two, causing the constexpr
-    # to be last. See the cpp_function_attributes test for a verification.
-    elif func.type.endswith('constexpr'):
-        func.type = func.type[:-10]
-        func.is_constexpr = True
-    else:
-        func.is_constexpr = False
-    # When 1.8.18 encounters `constexpr static`, it keeps the static there. For
-    # `static constexpr` it doesn't. In both cases the static="yes" is put
-    # there correctly. WHY DOXYGEN, WHY?!
-    if func.type.startswith('static'):
-        func.type = func.type[7:]
+    #
+    # First the prefix keywords - Doxygen has a habit of leaking attributes and
+    # other specifiers into the function's return type, and not necessarily
+    # in any consistent order (including swapping it with the actual type!)
+    func.is_constexpr = False
+    func.is_consteval = False
+    func.is_explicit = False
+    func.is_virtual = False
+    func.is_static = False
+    keywords_behaving_badly = (
+        'constexpr',
+        'consteval',
+        'explicit',
+        'static',
+        'friend',
+        'inline',
+        'virtual'
+    )
+    matched_bad_keyword = True
+    while matched_bad_keyword:
+        matched_bad_keyword = False
+        for kw in keywords_behaving_badly:
+            if func.type == kw: # constructors
+                matched_bad_keyword = True
+                func.type = ''
+            elif func.type.startswith(kw+' '):
+                matched_bad_keyword = True
+                func.type = func.type[len(kw):].strip()
+            elif func.type.endswith(' '+kw):
+                matched_bad_keyword = True
+                func.type = func.type[:len(kw)].strip()
+            else:
+                continue
+            if kw == 'constexpr':
+                func.is_constexpr = True
+            elif kw == 'consteval':
+                func.is_consteval = True
```

Comment:
This new template variable should ideally be listed also in `doc/documentation/doxygen.rst` somewhere down there in one of the tables.

Reply:
Yup, sure. Note that I've _also_ added `func.is_static`, which wasn't there previously either, just because it made the the logic consistent with the other members. I could have special-cased it with just a function-local `is_static` variable if you prefer.